// Copyright 2016 Google Inc. All Rights Reserved.
// Author: Sebastian Schaffert <schaffert@google.com>

syntax = "proto3";

package wikidata.primarysources.model;

// A message describing user statistics for an individual user.
message UserStatus {
    // User name.
    string name = 1;

    // Number of activities for this user.
    int64 activities = 2;
}

// A message describing current system runtime status;
message SystemStatus {
    // ISO GMT date when system was started.
    string startup = 1;

    // GIT version identifier of the current build.
    string version = 2;

    // Number of cache hits since system startup.
    int64 cache_hits = 3;

    // Number of cache misses since system startup.
    int64 cache_misses = 4;

    // Shared memory used by process (bytes).
    int64 shared_memory = 5;

    // Private memory used by process (bytes).
    int64 private_memory = 6;

    // Number of bytes used by process currently resident in memory.
    int64 resident_set_size = 7;
}

// A message describing current statement status in the database.
message StatementStatus {
    // Total number of statements.
    int64 statements = 1;

    // Number of approved statements.
    int64 approved = 2;

    // Number of unapproved statements.
    int64 unapproved = 3;

    // Number of duplicate statements.
    int64 duplicate = 4;

    // Number of blacklisted statements.
    int64 blacklisted = 5;

    // Number of wrong statements.
    int64 wrong = 6;
}

// Statistics about requests since system startup.
message RequestStatus {
    // Number of GET /entities/<id> requests since startup.
    int64 get_entity = 1;

    // Number of GET /entities/any requests since startup.
    int64 get_random = 2;

    // Number of GET /statements/<id> requests since startup.
    int64 get_statement = 3;

    // Number of POST /statements/<id> requests since startup.
    int64 update_statement = 4;

    // Number of GET /status requests since startup.
    int64 get_status = 5;
}

// A message describing the current status of the primary sources backend.
message Status {
    // Current statement statistics.
    StatementStatus statements = 1;

    // Current system statistics.
    SystemStatus system = 2;

    // Current request statistics.
    RequestStatus requests = 3;

    // User statistics for the top N users.
    repeated UserStatus top_users = 4;
}

